File: src/problems30_39/Problem30.java
/**
 * Surprisingly there are only three numbers that can be written as the sum of fourth powers of their digits:
 * <p/>
 * 1634 = 14 + 64 + 34 + 44
 * 8208 = 84 + 24 + 04 + 84
 * 9474 = 94 + 44 + 74 + 44
 * <p/>
 * As 1 = 14 is not a sum it is not included.
 * <p/>
 * The sum of these numbers is 1634 + 8208 + 9474 = 19316.
 * <p/>
 * Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.
 */
package problems30_39;

public class Problem30 {
    public static void main(String args[]) {
        int sum = 0;
        for (int x = 2; x < 200000; x++) {
            int intermediary = 0;
            String a = Integer.toString(x);
            for (int i = 0; i < a.length(); i++) {
                intermediary += Math.pow(Integer.parseInt(a.substring(i, i + 1)), 5);
            }
            if (intermediary == x) {
                sum += x;
                System.out.println(x);

            }
        }
        System.out.println(sum);
    }
}

File: problems30_39/Problem30.hs
-- log10(9^5 * 7) < 7 Therefore all numbers that can be written as the 5th power of their
-- digits are <= 1 000 000. 
-- Find the sum of the numbers that can be written as the sum of the fifth power of their
-- digits

pow5 x = x * x * x * x * x

pow5sum x =
	let ones = pow5 $ mod x 10
	    tens = pow5 $ mod (div x 10) 10
	    huns = pow5 $ mod (div x 100) 10
	    thou = pow5 $ mod (div x 1000) 10
	    ttho = pow5 $ mod (div x 10000) 10
	    htho = pow5 $ mod (div x 100000) 10
	in x == ones + tens + huns + thou + ttho + htho

main = do putStrLn $ show $ (sum [x | x <- [1..999999], pow5sum x]) - 1
